activerecord/lib/active_record/query_logs.rb

```diff
@@ -0,0 +1,187 @@
+  #
+  # * +controller+
+  # * +action+
+  # * +job+
+  #
+  # The tags used in a query can be configured directly:
+  #
+  #     ActiveRecord::QueryLogs.tags = [ :application, :controller, :action, :job ]
+  #
+  # or via Rails configuration:
+  #
+  #     config.active_record.query_log_tags = [ :application, :controller, :action, :job ]
+  #
+  # To add new comment tags, add a hash to the tags array containing the keys and values you
+  # want to add to the comment. Dynamic content can be created by setting a proc or lambda value in a hash,
+  # and can reference any value stored in the +context+ object.
+  #
+  # Example:
+  #
+  #    tags = [
+  #      :application,
+  #      { custom_tag: -> { context[:controller].controller_name } }
+  #    ]
+  #    ActiveRecord::QueryLogs.tags = tags
+  #
+  # The QueryLogs +context+ can be manipulated via +update_context+ & +set_context+ methods.
+  #
+  # Direct updates to a context value:
+  #
+  #    ActiveRecord::QueryLogs.update_context(foo: Bar.new)
+  #
+  # Temporary updates limited to the execution of a block:
+  #
+  #    ActiveRecord::QueryLogs.set_context(foo: Bar.new) do
+  #      posts = Post.all
+  #    end
+  #
+  # Tag comments can be prepended to the query:
+  #
+  #    ActiveRecord::QueryLogs.prepend_comment = true
+  #
+  # For applications where the content will not change during the lifetime of
+  # the request or job execution, the tags can be cached for reuse in every query:
+  #
+  #    ActiveRecord::QueryLogs.cache_query_log_tags = true
+  #
+  # This option can be set during application configuration or in a Rails initializer:
+  #
+  #    config.active_record.cache_query_log_tags = true
+  module QueryLogs
+    mattr_accessor :taggings, instance_accessor: false, default: {}
+    mattr_accessor :tags, instance_accessor: false, default: [ :application ]
+    mattr_accessor :prepend_comment, instance_accessor: false, default: false
+    mattr_accessor :cache_query_log_tags, instance_accessor: false, default: false
+    thread_mattr_accessor :cached_comment, instance_accessor: false
+
+    class << self
+      # Updates the context used to construct tags in the SQL comment.
+      # Resets the cached comment if <tt>cache_query_log_tags</tt> is +true+.
+      def update_context(**options)
+        context.merge!(**options.symbolize_keys)
+        self.cached_comment = nil
+      end
+
+      # Updates the context used to construct tags in the SQL comment during
+      # execution of the provided block. Resets provided values to nil after
+      # the block is executed.
+      def set_context(**options)
+        update_context(**options)
+        yield if block_given?
+      ensure
+        update_context(**options.transform_values! { nil })
+      end
+
+      # Temporarily tag any query executed within `&block`. Can be nested.
+      def with_tag(tag, &block)
+        self.inline_tags.push(tag)
+        yield if block_given?
+      ensure
+        self.inline_tags.pop
```

Comment:
Forgot to catch this, but we prefer implicit `self`.

```suggestion
        inline_tags.push(tag)
        yield if block_given?
      ensure
        inline_tags.pop
```